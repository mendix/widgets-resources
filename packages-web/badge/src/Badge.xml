<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.custom.badge.Badge" needsEntityContext="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../xsd/widget.xsd">
    <name>Badge</name>
    <description>Badges are numerical indicators of how many items are associated with a link</description>
    <icon>
        iVBORw0KGgoAAAANSUhEUgAAABAAAAAMCAYAAABr5z2BAAAAAXNSR0IArs4c6QAAASxJREFUKBWVUjFOxDAQ3LU3l0C4q2kRICFqmuMhUPMhPkBBeeILlAiJ5iRoQEh0FDzgIAh0WZvZGCMuBQpjWbtZz0x24zBNT9cOpvvn4436OGhLQ+C80OtbczG/eTgRqeLh1s7m0ePLO0V2FLEYy6KBsS2zmMGBaQ+au9v7M4nt51iV2YtRftMSnVGKcLDY4duthca0QhRixOF6iRSIYDPYFv+C6xxDFA+WBqJqZNlweO/IFN1r2fGPgXW6MnOvkMdx0Gg2ELgVPg3ZHyGPY3EV6VkUPiUG8UWfkOi5m/4X0aWdaxqhLD1JdwuomU9WGSfn2cE6wRytSwUhVSqKEdWV3f0w2DuaD7ChFX2+nl9dbs+KerIb22F/IovQslk8mdbMbE+wa+x/NQH+4gsBPW0QwdZQLAAAAABJRU5ErkJggg==
    </icon>
    <properties>
        <property key="label" type="translatableString" required="false">
            <caption>Label</caption>
            <category>Display</category>
            <description>The badge label</description>
            <translations>
                <translation lang="en_US"></translation>
                <translation lang="nl_NL"></translation>
            </translations>
        </property>
        <property key="badgeValue" type="translatableString" required="false">
            <caption>Value</caption>
            <category>Display</category>
            <description>The badge value</description>
            <translations>
                <translation lang="en_US"></translation>
                <translation lang="nl_NL"></translation>
            </translations>
        </property>
        <property key="bootstrapStyle" type="enumeration" defaultValue="default">
            <caption>Bootstrap style</caption>
            <category>Display</category>
            <description>The appearance of the badge based on Bootstrap classes</description>
            <enumerationValues>
                <enumerationValue key="default">Default</enumerationValue>
                <enumerationValue key="primary">Primary</enumerationValue>
                <enumerationValue key="success">Success</enumerationValue>
                <enumerationValue key="info">Info</enumerationValue>
                <enumerationValue key="warning">Warning</enumerationValue>
                <enumerationValue key="danger">Danger</enumerationValue>
            </enumerationValues>
        </property>
        <property key="badgeType" type="enumeration" required="true" defaultValue="badge">  
            <caption>Type</caption>
            <category>Display</category>
            <description>Render it as either a badge or a color label</description>
            <enumerationValues>
                <enumerationValue key="badge">Badge</enumerationValue>
                <enumerationValue key="label">Label</enumerationValue>
                </enumerationValues>
        </property>
        <property key="labelAttribute" type="attribute" required="false">
            <caption>Label attribute</caption>
            <category>Data source</category>
            <description>The attribute that contains the badge label</description>
            <attributeTypes>
                <attributeType name="Enum" />
                <attributeType name="String" />
            </attributeTypes>
        </property>
        <property key="valueAttribute" type="attribute" required="false">
            <caption>Value attribute</caption>
            <category>Data source</category>
            <description>The attribute that contains the badge value</description>
            <attributeTypes>
                <attributeType name="String" />
                <attributeType name="Integer" />
                <attributeType name="Decimal" />
                <attributeType name="Long" />
            </attributeTypes>
        </property>
        <property key="bootstrapStyleAttribute" type="attribute" required="false">
            <caption>Bootstrap style attribute</caption>
            <category>Data source</category>
            <description>The attribute value should be: default, primary, info, warning, danger, success.</description>
            <attributeTypes>
                <attributeType name="Enum" />
                <attributeType name="String" />
            </attributeTypes>
        </property>
        <property key="onClickEvent" type="enumeration" defaultValue="doNothing">
            <caption>On click</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="showPage">Show a page</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="microflow" type="microflow" required="false">
            <caption>Microflow</caption>
            <category>Events</category>
            <description>The microflow that will be executed on click</description>
            <returnType type="Void" />
        </property>
        <property key="page" type="form" required="false">
            <caption>Page</caption>
            <category>Events</category>
            <description>The page that will be opened on click</description>
        </property>
    </properties>
</widget>
