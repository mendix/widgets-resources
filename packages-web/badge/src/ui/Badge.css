.widget-badge-clickable {
    cursor: pointer;
}

.widget-badge {
    display: inline-block;
}

.widget-badge.badge:empty {
    display: initial;
    /* Fix padding to stay round */
    padding: 3px 10px;
}

.widget-badge.label:empty {
    display: initial;
    /* Fix padding to stay square */
    padding: .2em .8em .3em;
}

.widget-badge.badge {
    min-width: 18px;
    min-height: 18px;
}
