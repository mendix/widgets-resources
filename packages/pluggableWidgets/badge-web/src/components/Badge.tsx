import { createElement, CSSProperties, KeyboardEvent, ReactElement } from "react";
import classNames from "classnames";

export interface BadgeProps {
    type: "badge" | "label";
    className?: string;
    style?: CSSProperties;
    value: string;
    onClick?: () => void;
    onKeyDown?: (event: KeyboardEvent<HTMLSpanElement>) => void;
    tabIndex?: number;
}

export const Badge = (props: BadgeProps): ReactElement => {
    const { type, className, style, value, onClick, onKeyDown, tabIndex } = props;

    return (
        <span
            role={onClick || onKeyDown ? "button" : undefined}
            className={classNames("widget-badge", type, className, {
                "widget-badge-clickable": onClick
            })}
            onClick={onClick}
            onKeyDown={onKeyDown}
            style={style}
            tabIndex={tabIndex}
        >
            {value}
        </span>
    );
};
