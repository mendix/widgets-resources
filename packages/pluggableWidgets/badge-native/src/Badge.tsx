import { flattenStyles } from "@native-mobile-resources/util-widgets";
import { Component, createElement } from "react";
import { Platform, Text, TouchableNativeFeedback, TouchableOpacity, View } from "react-native";

import { BadgeProps } from "../typings/BadgeProps";
import { BadgeStyle, defaultBadgeStyle } from "./ui/Styles";
import { executeAction } from "@widgets-resources/piw-utils";

export type Props = BadgeProps<BadgeStyle>;

export class Badge extends Component<Props> {
    private readonly onClickHandler = this.onClick.bind(this);
    private readonly styles = flattenStyles(defaultBadgeStyle, this.props.style);

    render(): JSX.Element {
        const isAndroid = Platform.OS === "android";

        return (
            <View style={this.styles.container} testID={this.props.name}>
                {this.props.onClick ? (
                    isAndroid ? (
                        <TouchableNativeFeedback
                            background={TouchableNativeFeedback.SelectableBackgroundBorderless()}
                            onPress={this.onClickHandler}
                        >
                            {this.renderText()}
                        </TouchableNativeFeedback>
                    ) : (
                        <TouchableOpacity onPress={this.onClickHandler}>{this.renderText()}</TouchableOpacity>
                    )
                ) : (
                    this.renderText()
                )}
            </View>
        );
    }

    private renderText(): JSX.Element {
        const value = this.props.caption.value || "";

        return (
            <Text testID={`${this.props.name}$caption`} style={this.styles.caption}>
                {value}
            </Text>
        );
    }

    private onClick(): void {
        executeAction(this.props.onClick);
    }
}
