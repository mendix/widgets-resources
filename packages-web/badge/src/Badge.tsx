import { hot } from "react-hot-loader/root";
import { ReactNode, useCallback, createElement } from "react";
import { BadgeContainerProps } from "../typings/BadgeProps";
import { Badge as DisplayBadge } from "./components/Badge";
import { ValueStatus } from "mendix";
import { executeAction } from "@widgets-resources/piw-utils";

const Badge = (props: BadgeContainerProps): ReactNode => {
    const onClick = useCallback(() => {
        executeAction(props.onClick);
    }, [props.onClick]);

    return (
        <DisplayBadge
            type={props.type}
            defaultValue={
                props.defaultValue && props.defaultValue.status === ValueStatus.Available
                    ? props.defaultValue.value
                    : undefined
            }
            value={
                props.valueAttribute && props.valueAttribute.status === ValueStatus.Available
                    ? String(props.valueAttribute.value)
                    : undefined
            }
            bootstrapStyle={props.bootstrapStyle}
            clickable={props.onClick.canExecute}
            onClick={onClick}
            className={props.class}
            style={props.style}
        />
    );
};

export default hot(Badge);
