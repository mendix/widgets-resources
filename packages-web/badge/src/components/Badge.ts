import { SFC, createElement } from "react";
import * as classNames from "classnames";

import "../ui/Badge.css";

export interface BadgeProps {
    badgeType: "badge" | "label";
    defaultValue?: string;
    className?: string;
    style?: object;
    value?: string;
    bootstrapStyle?: BootstrapStyle;
    clickable?: boolean;
    onClickAction?: () => void;
    getRef?: (node: HTMLElement) => void;
}

export type BootstrapStyle = "default" | "info" | "inverse" | "primary" | "danger" | "success" | "warning";

export const Badge: SFC<BadgeProps> = (props) => createElement("span",
    {
        className: classNames("widget-badge", props.badgeType, {
            [`label-${props.bootstrapStyle}`]: !!props.bootstrapStyle,
            "widget-badge-clickable": props.clickable
        }),
        onClick: props.onClickAction,
        ref: props.getRef,
        style: props.style
    }, props.value || props.defaultValue);
