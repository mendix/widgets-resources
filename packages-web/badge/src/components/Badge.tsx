import { createElement, ReactElement } from "react";
import classNames from "classnames";

import "../ui/Badge.css";

export interface BadgeProps {
    type: "badge" | "label";
    defaultValue?: string;
    className?: string;
    style?: object;
    value?: string;
    bootstrapStyle?: BootstrapStyle;
    clickable?: boolean;
    onClick?: () => void;
    getRef?: (node: HTMLElement) => void;
}

export type BootstrapStyle = "default" | "info" | "inverse" | "primary" | "danger" | "success" | "warning";

export const Badge = (props: BadgeProps): ReactElement =>
    createElement(
        "span",
        {
            className: classNames("widget-badge", props.type, props.className, {
                [`label-${props.bootstrapStyle}`]: !!props.bootstrapStyle,
                "widget-badge-clickable": props.clickable
            }),
            onClick: props.onClick,
            ref: props.getRef,
            style: props.style
        },
        props.value || props.defaultValue
    );
