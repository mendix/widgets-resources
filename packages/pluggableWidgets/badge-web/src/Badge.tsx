import { ReactNode, useCallback, createElement } from "react";
import { executeAction } from "@widgets-resources/piw-utils";
import { ValueStatus } from "mendix";

import { BadgeContainerProps } from "../typings/BadgeProps";
import { Badge as DisplayBadge } from "./components/Badge";

import "./ui/BadgeContainer.css";

export default function Badge(props: BadgeContainerProps): ReactNode {
    const onClick = useCallback(() => {
        executeAction(props.onClick);
    }, [props.onClick]);

    return (
        <DisplayBadge
            type={props.type}
            value={props.value && props.value.status === ValueStatus.Available ? props.value.value : ""}
            brandStyle={props.brandStyle}
            clickable={props.onClick && props.onClick.canExecute}
            onClick={onClick}
            className={props.class}
            style={props.style}
        />
    );
}
