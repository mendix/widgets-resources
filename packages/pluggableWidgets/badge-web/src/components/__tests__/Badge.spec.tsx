import { shallow, ShallowWrapper } from "enzyme";
import { createElement } from "react";

import { Badge, BadgeProps } from "../Badge";

describe("Badge", () => {
    const createBadge = (props: BadgeProps): ShallowWrapper<any, any> => shallow(<Badge {...props} />);

    let defaultBadgeProps: BadgeProps;

    beforeEach(() => {
        defaultBadgeProps = {
            type: "badge",
            value: "text"
        };
    });

    it("renders as a badge", () => {
        const badge = createBadge(defaultBadgeProps);

        expect(badge).toMatchSnapshot();
    });

    it("renders as a label", () => {
        defaultBadgeProps.type = "label";
        const badge = createBadge(defaultBadgeProps);

        expect(badge).toMatchSnapshot();
    });

    it("renders when an empty string is passed as value", () => {
        defaultBadgeProps.value = "";
        const badge = createBadge(defaultBadgeProps);

        expect(badge).toMatchSnapshot();
    });

    it("renders as a button like element when onClick function is passed", () => {
        defaultBadgeProps.onClick = jest.fn();
        const badge = createBadge(defaultBadgeProps);

        expect(badge).toMatchSnapshot();
    });

    it("triggers onClick function with a click event", () => {
        defaultBadgeProps.onClick = jest.fn();
        const badge = createBadge(defaultBadgeProps);

        badge.simulate("click");

        expect(defaultBadgeProps.onClick).toHaveBeenCalledTimes(1);
    });

    it("renders as a button like element when onKeyDown function is passed", () => {
        defaultBadgeProps.onKeyDown = jest.fn();
        const badge = createBadge(defaultBadgeProps);

        expect(badge).toMatchSnapshot();
    });

    it("triggers onKeyDown function on key down", () => {
        defaultBadgeProps.onKeyDown = jest.fn();
        const badge = createBadge(defaultBadgeProps);

        badge.simulate("keydown");

        expect(defaultBadgeProps.onKeyDown).toHaveBeenCalledTimes(1);
    });

    it("renders with a tabIndex", () => {
        defaultBadgeProps.tabIndex = 1;
        const badge = createBadge(defaultBadgeProps);

        expect(badge).toMatchSnapshot();
    });

    it("renders custom classes", () => {
        defaultBadgeProps.className = "custom-class";
        const badge = createBadge(defaultBadgeProps);

        expect(badge).toMatchSnapshot();
    });

    it("renders custom styles", () => {
        defaultBadgeProps.style = { padding: 5 };
        const badge = createBadge(defaultBadgeProps);

        expect(badge).toMatchSnapshot();
    });
});
