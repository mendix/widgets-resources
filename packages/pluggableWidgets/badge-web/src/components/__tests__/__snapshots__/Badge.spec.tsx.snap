// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders as a badge 1`] = `
<span
  className="widget-badge badge"
>
  text
</span>
`;

exports[`Badge renders as a button like element when onClick function is passed 1`] = `
<span
  className="widget-badge badge widget-badge-clickable"
  onClick={[MockFunction]}
  role="button"
>
  text
</span>
`;

exports[`Badge renders as a button like element when onKeyDown function is passed 1`] = `
<span
  className="widget-badge badge"
  onKeyDown={[MockFunction]}
  role="button"
>
  text
</span>
`;

exports[`Badge renders as a label 1`] = `
<span
  className="widget-badge label"
>
  text
</span>
`;

exports[`Badge renders custom classes 1`] = `
<span
  className="widget-badge badge custom-class"
>
  text
</span>
`;

exports[`Badge renders custom styles 1`] = `
<span
  className="widget-badge badge"
  style={
    Object {
      "padding": 5,
    }
  }
>
  text
</span>
`;

exports[`Badge renders when an empty string is passed as value 1`] = `
<span
  className="widget-badge badge"
/>
`;

exports[`Badge renders with a tabIndex 1`] = `
<span
  className="widget-badge badge"
  tabIndex={1}
>
  text
</span>
`;
